### DISCUSSION TITLE
Is there a way to rollback to (or reinstall) specific controller versions?

### DISCUSSION THREAD

**Discussion by mechanicalgirl** [fluxcd/flux2#4578] [General]
I know how to rollback a system state, but I'm looking for a way to rollback to specific versions of Flux and the source/notification/helm/etc. controllers.

I was just assigned to a project that's been sadly neglected for a while, and the flux versions are out of date. I'm trying to schedule a time to upgrade, but the dev team doesn't want it done without a plan for reverting if something goes wrong. Reasonable, sure, but I can't find anything in the flux upgrade docs that addresses how to, well, _downgrade_ if necessary.

I've been using bootstrap to upgrade on test clusters - if I specify a toolkit version with `--version` will that install specific corresponding controller versions? Can I just use bootstrap to downgrade that way, by specifying an older version? Or, if I have to uninstall altogether, can I reinstall and pin controllers to older versions? Is it possible to pin older versions of controllers as patches in the kustomization chart? 

Is there a best practice for controller rollbacks/version pinning?

**Comment by stefanprodan** [comment]
To rollback to any version:
- Uninstall with `flux uninstall --keep-namespace` (this will not affect any deployed workloads)
- Download an older CLI version from https://github.com/fluxcd/flux2/releases
- Bootstrap using the same command args with the downloaded CLI

Note that CLI versions are not backwards compatible between minor versions, so you need to match the CLI version with the Flux control plane.
